### PR TITLE
tweak registry remapping logic

### PR DESCRIPTION
**Pull request by asiekierka**

### CODE REVIEW

File: src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java [hunk 1]
```diff
@@ -79,87 +79,120 @@ public void setPre(int id, Identifier identifier, Object object, CallbackInfoRet
...
+			int id = remoteIndexedEntries.getInt(identifier);
 			T object = entries.get(identifier);
+
+			// Warn if an object is missing from the local registry.
+			// This should only happen in AUTHORITATIVE mode, and as such we
+			// throw an exception otherwise.
 			if (object == null) {
-				LOGGER.warn(identifier + " missing from registry, but requested!");
+				if (mode != RemapMode.AUTHORITATIVE) {
+					throw new RemapException(identifier + " missing from registry, but requested!");
```
Comment (Boundarybreaker) [inline src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java]: Does this mean that trying to connect to a server that has blocks/items/etc. that the client doesn't will throw?

File: src/main/java/net/fabricmc/fabric/impl/registry/RemappableRegistry.java [hunk 2]
```diff
@@ -20,6 +20,12 @@
 import net.minecraft.util.Identifier;
 
 public interface RemappableRegistry {
-	void remap(Object2IntMap<Identifier> idMap, boolean reallocateMissingEntries) throws RemapException;
+	enum RemapMode {
+		AUTHORITATIVE, /* remapping the server; any changes allowed */
```
Comment (InsomniaKitten) [inline src/main/java/net/fabricmc/fabric/impl/registry/RemappableRegistry.java]: These should be javadocs, not comment blocks

File: src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java [hunk 3]
```diff
@@ -79,87 +79,120 @@ public void setPre(int id, Identifier identifier, Object object, CallbackInfoRet
...
 
-		Object defaultValue = null;
-		//noinspection ConstantConditions
-		if (registry instanceof DefaultedRegistry) {
-			defaultValue = registry.get(((DefaultedRegistry<Object>) registry).getDefaultId());
-		}
-
-		if (!reallocateMissingEntries && !idMap.keySet().equals(registry.getIds())) {
-			throw new RemapException("Source and destination keys differ!");
+		// Throw on invalid conditions.
```
Comment (InsomniaKitten) [inline src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java]: This should be a javadoc on the encompassing method, describing the behaviour

File: src/main/java/net/fabricmc/fabric/impl/registry/RegistrySyncManager.java [hunk 4]
```diff
@@ -51,24 +52,20 @@ private RegistrySyncManager() {
...
-						// TODO: log error properly
-						e.printStackTrace();
+						errorHandler.accept(e);
 					}
 				}).get(30, TimeUnit.SECONDS);
-			} catch (ExecutionException e) {
-				e.printStackTrace();
-			} catch (InterruptedException | TimeoutException e) {
-				// TODO: better error handling
-				new Exception("Failed to apply received packets in time!", e).printStackTrace();
```
Comment (InsomniaKitten) [inline src/main/java/net/fabricmc/fabric/impl/registry/RegistrySyncManager.java]: This should be printed with a logger

File: src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java (hunk 1 above)
Comment (asiekierka) [reply src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java]: Yes.

File: src/main/java/net/fabricmc/fabric/impl/registry/RemappableRegistry.java (hunk 2 above)
Comment (asiekierka) [reply src/main/java/net/fabricmc/fabric/impl/registry/RemappableRegistry.java]: This is internal code for the time being

File: src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java (hunk 3 above)
Comment (asiekierka) [reply src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java]: This is *super* internal code. If you're an user looking at it, you've done many things wrong.

**Comment by asiekierka** [comment]
It is functionally complete now - disconnects work as intended.

**Pull request closed without merging**